exasol/driver/websocket/_cursor.py

```
@@ -0,0 +1,284 @@
+    display_size: Optional[int] = None
+    internal_size: Optional[int] = None
+    precision: Optional[int] = None
+    scale: Optional[int] = None
+    null_ok: Optional[bool] = None
+
+
+def _metadata(name, metadata) -> MetaData:
+    type_mapping = {t.value: t for t in TypeCode}
+    key_mapping = {
+        "name": "name",
+        "type_code": "type",
+        "precision": "precision",
+        "scale": "scale",
+        "display_size": "unknown",
+        "internal_size": "size",
+        "null_ok": "unknown",
+    }
+    metadata = defaultdict(lambda: None, metadata)
+    metadata["type"] = type_mapping[metadata["type"]]
+    metadata["name"] = name
+    return MetaData(**{k: metadata[v] for k, v in key_mapping.items()})
+
+
+def _is_not_closed(method):
+    """
+    Decorator requires the object to have a result.
+
+    Raises:
+        Error if the cursor object has not produced a result yet.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if self._is_closed:
+            raise Error("Cursor is closed, further operations are not permitted.")
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+def _requires_result(method):
+    """
+    Decorator requires the object to have a result.
+
+    Raises:
+        Error if the cursor object has not produced a result yet.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if not self._cursor:
+            raise Error("No result has been produced.")
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+def _identity(value):
+    return value
+
+
+def _pyexasol2dbapi(value, metadata):
+    members = (
+        "name",
+        "type_code",
+        "display_size",
+        "internal_size",
+        "precision",
+        "scale",
+        "null_ok",
+    )
+    metadata = MetaData(**{k: v for k, v in zip(members, metadata)})
+
+    def to_date(v):
+        if not isinstance(v, str):
+            return v
```

Shouldn't this check that the instance is a date before returning without change?

What if someone supplies a number.

---

then datetime.date.fromisoformat(v) would fail and this is probably a case where there is no other option then fail hard, because we got something from the DB which is garbage

---

As discussed this is a function scoped in a private function, my call sites are expected to be correct.